src/DurableTask.AzureStorage/AzureStorageOrchestrationService.cs

```diff
@@ -1215,11 +1215,15 @@ await this.CommitOutboundQueueMessages(
                     this.orchestrationSessionManager.AddMessageToPendingOrchestration(session.ControlQueue, messages, session.TraceActivityId, CancellationToken.None);
                 }
             }
-            catch (RequestFailedException rfe) when (rfe.Status == (int)HttpStatusCode.PreconditionFailed)
+            // If this is the first work item for the orchestration and it has already been completed, then session.Etag will be null and a new history will attempt to be inserted when one already exists.
+            // In this case, the status code will be "Conflict".
+            // If this is not the first work item and it has already been completed, then session.Etag will be non-null but will not match the etag stored when attempting to update the history.
+            // In this case, the status code will be "PreconditionFailed".
+            catch (DurableTaskStorageException dtse) when (dtse.HttpStatusCode == (int)HttpStatusCode.Conflict || dtse.HttpStatusCode == (int)HttpStatusCode.PreconditionFailed)
             {
                 // Precondition failure is expected to be handled internally and logged as a warning.
                 // The orchestration dispatcher will handle this exception by abandoning the work item
-                throw new SessionAbortedException("Aborting execution due to failed precondition.", rfe);
+                throw new SessionAbortedException("Aborting execution due to conflicting completion of the work item.", dtse);
```

Comment:
nit - if this is possibly customer facing, it's useful to add a little hint here that the error is benign. Consider:

```suggestion
                throw new SessionAbortedException("Aborting execution because the work item was already completed. This can occur when another worker took over processing, and is benign when it occurs rarely", dtse);

```

Reply:
I _think_ `SessionAbortedExceptions` do not surface, though someone else from the team correct me if I'm wrong, please.

Reply:
I won't insist on this one, it's my personal style to augment these nonetheless but it really doesn't matter. Feel free to dismiss.